timemachine/cpp/src/rmsd_restraint.cu

```diff
@@ -173,52 +196,41 @@ void RMSDRestraint<RealType>::execute_device(
-    Eigen::MatrixXd x2_adjoint = (c_adjoint*x1.transpose()).transpose();
-    Eigen::MatrixXd x1_adjoint = x2*c_adjoint;
-
-    std::vector<unsigned long long> du_dx(N*3, 0);
-
-    for(int b=0; b < B; b++) {
-        int src_idx = h_atom_map_[b*2+0];
-        du_dx[src_idx*3+0] += static_cast<unsigned long long>(llrint(k_*x1_adjoint(b, 0)*FIXED_EXPONENT));
-        du_dx[src_idx*3+1] += static_cast<unsigned long long>(llrint(k_*x1_adjoint(b, 1)*FIXED_EXPONENT));
-        du_dx[src_idx*3+2] += static_cast<unsigned long long>(llrint(k_*x1_adjoint(b, 2)*FIXED_EXPONENT));
-
-        int dst_idx = h_atom_map_[b*2+1];
-        du_dx[dst_idx*3+0] += static_cast<unsigned long long>(llrint(k_*x2_adjoint(b, 0)*FIXED_EXPONENT));
-        du_dx[dst_idx*3+1] += static_cast<unsigned long long>(llrint(k_*x2_adjoint(b, 1)*FIXED_EXPONENT));
-        du_dx[dst_idx*3+2] += static_cast<unsigned long long>(llrint(k_*x2_adjoint(b, 2)*FIXED_EXPONENT));
-    }
-
-    gpuErrchk(cudaMemcpy(d_u_buf_, &nrg, sizeof(*d_u_buf_), cudaMemcpyHostToDevice));
-    gpuErrchk(cudaMemcpy(d_du_dx_buf_, &du_dx[0], N*3*sizeof(*d_du_dx_buf_), cudaMemcpyHostToDevice));
-
-    const int blocks= (N_+32-1)/32;
-    const int tpb = 32;
-
-    k_finalize<<<blocks, tpb, 0, stream>>>(
-        N,
-        d_du_dx_buf_,
-        d_u_buf_,
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> v_a = v_a_t.transpose();
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> lhs_1 = u*F.cwiseProduct(u.transpose()*u_a - u_a.transpose()*u)*smat;
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> lhs_2 = (eye - u*u.transpose())*u_a*smat_inv;
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> lhs = (lhs_1+lhs_2)*v.transpose();
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> rhs_1 = smat*(F.cwiseProduct(v.transpose()*v_a - v_a.transpose()*v))*v.transpose();
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> rhs_2 = smat_inv*v_a.transpose()*(eye - v*v.transpose());
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> rhs = u*(rhs_1 + rhs_2);
+    Eigen::Matrix<RealType, Eigen::Dynamic, Eigen::Dynamic> c_adjoint = (lhs + rhs).cast<RealType>();
+    // Eigen::RowMajor required to be able to copy out of device
+    Eigen::Matrix<RealType, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> x2_adjoint = (c_adjoint*x1.transpose()).transpose();
+    Eigen::Matrix<RealType, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> x1_adjoint = x2*c_adjoint;
+
+    // Reusing d_centroid_a_ and d_centroid_b_ for x1 and x2 adjoints
+    gpuErrchk(cudaMemcpy(d_centroid_a_, x1_adjoint.data(), B*3*sizeof(RealType), cudaMemcpyHostToDevice));
+    gpuErrchk(cudaMemcpy(d_centroid_b_, x2_adjoint.data(), B*3*sizeof(RealType), cudaMemcpyHostToDevice));
+
+    RealType squared_term = term*term;
+
+    k_finalize<<<grid_blocks, tpb, 0, stream>>>(
+        d_atom_map_,
+        d_centroid_a_,
+        d_centroid_b_,
+        B,
+        k_,
+        squared_term,
```

Comment:
Rather than passing a buffer, i pass the RealType directly. Seemed cleaner, though perhaps there is a performance impact?